ibtk/src/math/PETScMatUtilities.cpp

```diff
@@ -806,10 +618,460 @@ void PETScMatUtilities::constructPatchLevelSCInterpOp(Mat& mat,
+                        PETSC_DEFAULT, &o_nnz[0], &mat);
+    IBTK_CHKERRQ(ierr);
+
+// Set some general matrix options.
+#if !defined(NDEBUG)
+    ierr = MatSetOption(mat, MAT_NEW_NONZERO_LOCATION_ERR, PETSC_TRUE);
+    IBTK_CHKERRQ(ierr);
+    ierr = MatSetOption(mat, MAT_NEW_NONZERO_ALLOCATION_ERR, PETSC_TRUE);
+    IBTK_CHKERRQ(ierr);
+#endif
+
+    // Determine the matrix-coefficients
+    for (PatchLevel<NDIM>::Iterator p(fine_patch_level); p; p++)
+    {
+        Pointer<Patch<NDIM> > fine_patch = fine_patch_level->getPatch(p());
+        const Box<NDIM>& fine_patch_box = fine_patch->getBox();
+        Pointer<SideData<NDIM, int> > fine_dof_data = fine_patch->getPatchData(dof_index_idx);
+        const unsigned depth = fine_dof_data->getDepth();
+        std::vector<int> samrai_petsc_map(2 * depth);
+
+        for (int axis = 0; axis < NDIM; ++axis)
+        {
+            IntVector<NDIM> offset = 0;
+            offset(axis) = 1;
+
+            int data_offset = 0;
+            for (int side = 0; side < axis; ++side)
+            {
+                int side_offset = depth;
+                for (unsigned d = 0; d < NDIM; ++d) side_offset *= coarse_num_cells[side](d);
+                data_offset += side_offset;
+            }
+
+            for (Box<NDIM>::Iterator b(SideGeometry<NDIM>::toSideBox(fine_patch_box, axis)); b; b++)
+            {
+                const CellIndex<NDIM>& i = b();
+                const SideIndex<NDIM> i_s(i, axis, SideIndex<NDIM>::Lower);
+                bool on_proc_fine_loc = true;
+                for (unsigned d = 0; d < depth; ++d)
+                {
+                    const int fine_dof_idx = (*fine_dof_data)(i_s, d);
+                    on_proc_fine_loc = on_proc_fine_loc && fine_dof_idx >= i_fine_lower && fine_dof_idx < i_fine_upper;
+                }
+                if (!on_proc_fine_loc) continue;
+
+                const CellIndex<NDIM> I = IndexUtilities::coarsen(i, fine_coarse_ratio);
+                const CellIndex<NDIM>& I_L = I;
+                const CellIndex<NDIM> I_U = I_L + offset;
+                for (unsigned d = 0; d < depth; ++d)
+                {
+                    samrai_petsc_map[d] = IndexUtilities::getIntegerMapping(
+                        I_L, coarse_domain_lower, coarse_num_cells[axis], d, coarse_ao_offset + data_offset);
+                    samrai_petsc_map[depth + d] = IndexUtilities::getIntegerMapping(
+                        I_U, coarse_domain_lower, coarse_num_cells[axis], d, coarse_ao_offset + data_offset);
+                }
+                AOApplicationToPetsc(coarse_level_ao, 2 * depth, &samrai_petsc_map[0]);
+
+                for (unsigned d = 0; d < depth; ++d)
+                {
+                    int row = (*fine_dof_data)(i_s, d);
+                    int col[2] = { samrai_petsc_map[d], samrai_petsc_map[depth + d] };
+
+                    // To avoid calling refine(), the term in the bracket is same as:
```

Comment:
> On Apr 23, 2015, at 10:20 PM, Amneet Bhalla <notifications@github.com> wrote:
> 
> In ibtk/src/math/PETScMatUtilities.cpp <https://github.com/IBAMR/IBAMR/pull/6#discussion_r29019887>:
> 
> > +                const CellIndex<NDIM> I_U = I_L + offset;
> > +                for (unsigned d = 0; d < depth; ++d)
> > +                {
> > +                    samrai_petsc_map[d] = IndexUtilities::getIntegerMapping(
> > +                        I_L, coarse_domain_lower, coarse_num_cells[axis], d, coarse_ao_offset + data_offset);
> > +                    samrai_petsc_map[depth + d] = IndexUtilities::getIntegerMapping(
> > +                        I_U, coarse_domain_lower, coarse_num_cells[axis], d, coarse_ao_offset + data_offset);
> > +                }
> > +                AOApplicationToPetsc(coarse_level_ao, 2 * depth, &samrai_petsc_map[0]);
> > +
> > +                for (unsigned d = 0; d < depth; ++d)
> > +                {
> > +                    int row = (*fine_dof_data)(i_s, d);
> > +                    int col[2] = { samrai_petsc_map[d], samrai_petsc_map[depth + d] };
> > +
> > +                    // To avoid calling refine(), the term in the bracket is same as:
> Wee-bee of optimization ;-). Is the present formula confusing?
> 

https://shreevatsa.wordpress.com/2008/05/16/premature-optimization-is-the-root-of-all-evil/ <https://shreevatsa.wordpress.com/2008/05/16/premature-optimization-is-the-root-of-all-evil/>